### PR TITLE
Added validation for upsert request

### PR DESCRIPTION
**Pull request by kunal642**
related to #16671

### CODE REVIEW

**Comment by kunal642** [comment]
Can someone please review this

File: core/src/test/java/org/elasticsearch/action/update/UpdateRequestTests.java [hunk 1]
```diff
@@ -485,4 +486,19 @@ public void testToAndFromXContent() throws IOException {
         BytesReference finalBytes = toXContent(parsedUpdateRequest, xContentType, humanReadable);
         assertToXContentEquivalent(originalBytes, finalBytes, xContentType);
     }
+
+    public void testToValidateUpsertRequestAndVersion() {
+        UpdateRequest updateRequest = new UpdateRequest();
+        updateRequest.version(1L);
+        updateRequest.upsert(new IndexRequest("index","type", "1"));
+        assertThat(updateRequest.validate().validationErrors(), contains("index is missing",
+            "can't provide both upsert request and a version","type is missing","id is missing","script or doc is missing"));
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/action/update/UpdateRequestTests.java]: Can you set the missing things so they don't throw as well?

**Review by nik9000 — CHANGES_REQUESTED** [review]

**Comment by nik9000** [comment]
I think it is worth also writing a test for bulk requests.  They can attempt to set the version on an upsert request.

**Comment by kunal642** [comment]
@nik9000 the requested changes have been done

File: core/src/test/java/org/elasticsearch/action/bulk/BulkRequestTests.java [hunk 2]
```diff
@@ -255,4 +255,24 @@ public void testSmileIsSupported() throws IOException {
         assertEquals(1, request.sourceAsMap().size());
         assertEquals("value", request.sourceAsMap().get("field"));
     }
+
+    public void testToValidateUpsertRequestAndVersionInBulkRequest() {
+        UpdateRequest updateRequest = new UpdateRequest("index", "type", "id");
+        updateRequest.version(1L);
+        updateRequest.doc("{}", XContentType.JSON);
+        updateRequest.upsert(new IndexRequest("index","type", "id").source("{}", XContentType.JSON));
+        BulkRequest bulkRequest = new BulkRequest();
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/action/bulk/BulkRequestTests.java]: I meant a test that uses the bulk request's overrides in `BulkRequest#add`. That way we are testing a way of "sneaking" the version into the request that is unique to bulk requests.

**Review by nik9000 — CHANGES_REQUESTED** [review]

**Comment by kunal642** [comment]
@nik9000 changed the test case

**Comment by nik9000** [comment]
> @nik9000 changed the test case

Thanks! I've spent all day merging things and haven't gotten around to looking at this yet. I promise I'll get to it soon!

**Review by nik9000 — APPROVED** [review]
Actually, I did get a chance to look at it. That is perfect. I'll have a look at merging tomorrow.

**Review by nik9000 — CHANGES_REQUESTED** [review]
I hate to do this after approving the change but running `gradle check` caught this:
https://github.com/elastic/elasticsearch/blob/master/core/src/test/java/org/elasticsearch/update/UpdateIT.java#L535-L543

That test is now out of date. Can you remove it entirely with your PR?

You might have to merge master into your PR if you ran tests locally and didn't catch it. It is what I did before running the tests.

**Comment by kunal642** [comment]
@nik9000 i dont know how i missed it
Ill remove the marked test case

**Comment by kunal642** [comment]
@nik9000 i tried running gradle test and all the test suites related to plugins are failing.

**Comment by nik9000** [comment]
@kunal642 to be honest I never got that far. `gradle check` stopped me in core before I got to plugins.

**Comment by kunal642** [comment]
@nik9000 should i push??

**Comment by nik9000** [comment]
> @nik9000 should i push??

Sure! If you push what you have I can have a look at the failure and advise from there.

**Comment by kunal642** [comment]
@nik9000 please have a look

**Comment by nik9000** [comment]
So far all the tests seem fine to me....

**Comment by kunal642** [comment]
Thats great
I guess there is some problem in my repo

When i normally run the test cases they give me jar hell exception
In debug mode they run fine

Gradle check stalls the test cases for seconds and gives timeout exception after a while

**Review by nik9000 — APPROVED** [review]

**Comment by nik9000** [comment]
@elasticmachine, this PR is merged.

**Pull request merged**